be/src/vec/functions/array/function_array_sort.cpp

```diff
@@ -15,12 +15,203 @@
+    }
+
+    static bool _execute_string(const IColumn& src_column, const ColumnArray::Offsets& src_offsets,
+                                IColumn& dest_column, ColumnArray::Offsets& dest_offsets,
+                                const NullMapType* src_null_map, NullMapType* dest_null_map) {
+        const ColumnString* src_data_concrete = reinterpret_cast<const ColumnString*>(&src_column);
+        if (!src_data_concrete) {
+            return false;
+        }
+
+        ColumnString& dest_column_string = reinterpret_cast<ColumnString&>(dest_column);
+        ColumnString::Chars& column_string_chars = dest_column_string.get_chars();
+        ColumnString::Offsets& column_string_offsets = dest_column_string.get_offsets();
+        column_string_chars.reserve(src_column.size());
+
+        ColumnArray::Offset prev_src_offset = 0;
+        IColumn::Permutation permutation(src_column.size());
+        for (size_t i = 0; i < src_column.size(); ++i) {
+            permutation[i] = i;
+        }
+
+        for (auto curr_src_offset : src_offsets) {
+            // filter and insert null element first
+            for (ColumnArray::Offset j = prev_src_offset; j < curr_src_offset; ++j) {
+                if (src_null_map && (*src_null_map)[j]) {
+                    DCHECK(dest_null_map != nullptr);
+                    column_string_offsets.push_back(column_string_offsets.back());
+                    (*dest_null_map).push_back(true);
+                }
+            }
+
+            _sort_by_permutation<ColumnString>(prev_src_offset, curr_src_offset, src_data_concrete,
+                                               src_column, src_null_map, permutation);
+
+            // insert non-null element after sort by permutation
+            for (ColumnArray::Offset j = prev_src_offset; j < curr_src_offset; ++j) {
+                if (src_null_map && (*src_null_map)[j]) {
+                    continue;
+                }
+
+                StringRef src_str_ref = src_data_concrete->get_data_at(permutation[j]);
+                // copy the src data to column_string_chars
+                const size_t old_size = column_string_chars.size();
+                const size_t new_size = old_size + src_str_ref.size + 1;
+                column_string_chars.resize(new_size);
+                if (src_str_ref.size > 0) {
+                    memcpy(column_string_chars.data() + old_size, src_str_ref.data,
+                           src_str_ref.size);
+                }
+                column_string_chars[old_size + src_str_ref.size] = 0;
+                column_string_offsets.push_back(new_size);
+
+                if (dest_null_map) {
+                    (*dest_null_map).push_back(false);
+                }
+            }
+            dest_offsets.push_back(curr_src_offset);
+            prev_src_offset = curr_src_offset;
+        }
+        return true;
+    }
+
+    static bool _execute_by_type(const IColumn& src_column, const ColumnArray::Offsets& src_offsets,
```

Comment:
This function is completely duplicate with the one in function_array_distinct.cpp.
You need to think for an abstract method or impl.

Reply:
done